<a href="https://colab.research.google.com/github/tinamilo6/Textmining/blob/main/A03_TextMining1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pandas sklearn nltk

  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


# Assignment 3 - Text Mining

Project management and tools for health informatics

## 1. Download and prepare data:

**Do not alter the code in this Section!**

The code in this section downloads the [IMDB IMDB Large Movie Review Dataset]('https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz') which is the dataset you will be working on in this assignment.

In [2]:
import os
import tarfile
from urllib.request import urlretrieve

In [3]:
if not os.path.exists('aclImdb'):
    # download data:
    urlretrieve('https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz', 'aclImdb.tar.gz')

    # unzip data:
    with tarfile.open('aclImdb.tar.gz') as file:
        file.extractall('./')

## 2. Some helper Functions:

**Do not alter the code in this Section!**

This section contains the code for some helper functions that will be useful for solving the assignment. Example code on how to use the functions is provided in section 3.

In [4]:
import pandas as pd

from typing import Literal, Tuple, Iterable

Function for loading data into a pandas dataframe:

In [5]:
def load_data(split:Literal['train', 'test'], texts_per_class:int=500) -> pd.DataFrame:
    ''' Loads the data into a pandas dataframe.'''
    paths  = []
    labels = []

    for label in ('pos', 'neg'):
        # get all files in the folder:
        files = os.listdir(os.path.join('aclImdb', split, label))[:texts_per_class]

        # append them to the lists:
        paths.extend([os.path.join('aclImdb', split, label, f) for f in files])
        labels.extend([label] * len(files))

    return pd.DataFrame({'path':paths, 'label':labels})

Function for loading a specific text:

In [6]:
def load_text(path:str) -> str:
    ''' Reads a single text given the path. '''
    # read file from disk:
    with open(path, 'r', encoding='utf8') as file:
        s = file.read()

    return s

Function for iterating through multiple texts:

In [7]:
def iterate_texts(data:pd.DataFrame) -> Iterable[Tuple[str, str]]:
    ''' Iterates through a pandas dataframe. '''

    for path in data['path'].values:
        # read file from disk:
        with open(path, 'r', encoding='utf8') as file:
            text = file.read()

        yield text

## 3. Your Code:

**Alter the code below to complete the assignment!**

Load the training data:

In [8]:
data_train = load_data('train')
data_test  = load_data('test')
data_train

path label
0    aclImdb/train/pos/11266_10.txt   pos
1     aclImdb/train/pos/11392_8.txt   pos
2    aclImdb/train/pos/10098_10.txt   pos
3      aclImdb/train/pos/1243_9.txt   pos
4     aclImdb/train/pos/3890_10.txt   pos
..                              ...   ...
995    aclImdb/train/neg/8531_3.txt   neg
996   aclImdb/train/neg/10706_4.txt   neg
997    aclImdb/train/neg/8834_4.txt   neg
998    aclImdb/train/neg/3574_1.txt   neg
999    aclImdb/train/neg/6924_1.txt   neg

[1000 rows x 2 columns]

### Accessing the texts:

In [9]:
# Sample code: load a single text
load_text(data_train.loc[0, 'path'])

'I recently watched this, but when it started I had no idea what the concept was about, what the topic was.....in short - I had no idea what it was. Was it a documentary, was it a comedy routine.....Well, it was BOTH.<br /><br />It started a little slow, but I think that\'s because I had absolutely no idea what type of program I was viewing. But it quickly sucked me in. The episode I watched had Robert Wuhl discussing fact and fiction in history. Mainly how we (american\'s) learn history that isn\'t really true - and how we got to learn what we did. He did this in such a way as to keep the viewer completely entertained, and interested. I actually learned a few things and that is a true indicator of how effective this type of program can be.<br /><br />I would love to see this picked up as a series for HBO. I believe it can be just as fun and effective with a variety of topics - especially if they are "taught" in the same type of manner as this episode.'

In [10]:
# Sample code: iterate through all texts
for text in iterate_texts(data_train[:20]):
    print(text)

I recently watched this, but when it started I had no idea what the concept was about, what the topic was.....in short - I had no idea what it was. Was it a documentary, was it a comedy routine.....Well, it was BOTH.<br /><br />It started a little slow, but I think that's because I had absolutely no idea what type of program I was viewing. But it quickly sucked me in. The episode I watched had Robert Wuhl discussing fact and fiction in history. Mainly how we (american's) learn history that isn't really true - and how we got to learn what we did. He did this in such a way as to keep the viewer completely entertained, and interested. I actually learned a few things and that is a true indicator of how effective this type of program can be.<br /><br />I would love to see this picked up as a series for HBO. I believe it can be just as fun and effective with a variety of topics - especially if they are "taught" in the same type of manner as this episode.
Paris is the place to be to enjoy bea

In [11]:
# Define the number of sample texts to print
n_samples = 7

# Assume 'data_train' is a DataFrame or list containing your text data.
# Instead of accessing 'text', access the 'path' column,
# which likely contains the file paths to your text data.
# Then, load the text content from those paths.
X_train = [load_text(data_train.loc[i, 'path']) for i in range(n_samples)]

# Print some sample texts from the training set
for i in range(n_samples):
    print(f"Sample {i+1}:\n")
    print(X_train[i])  # Print the text
    print("\n" + "="*80 + "\n")

Sample 1:

I recently watched this, but when it started I had no idea what the concept was about, what the topic was.....in short - I had no idea what it was. Was it a documentary, was it a comedy routine.....Well, it was BOTH.<br /><br />It started a little slow, but I think that's because I had absolutely no idea what type of program I was viewing. But it quickly sucked me in. The episode I watched had Robert Wuhl discussing fact and fiction in history. Mainly how we (american's) learn history that isn't really true - and how we got to learn what we did. He did this in such a way as to keep the viewer completely entertained, and interested. I actually learned a few things and that is a true indicator of how effective this type of program can be.<br /><br />I would love to see this picked up as a series for HBO. I believe it can be just as fun and effective with a variety of topics - especially if they are "taught" in the same type of manner as this episode.


Sample 2:

Paris is the 

LIMITATIONS/ISSUES CAPTURED FROM THE ABOVE TEXT SAMPLES:

HTML tags and special characters

Punctuation and symbols (&)

Contractions ("isn't", "I'll", "I'm")

Parentheses and annotations ("(Crouching Tiger) ")

Informal formatting ("my rating is ****")

Ambiguity and polysemy ("dictators")

Long and complexes sentences

Informal language ("what can be so bad about that?")

Quotation marks ("dictators", "sin")



We will elaborate about them and add our conclusions in the report.

### A simple pipeline:

**White-Space tokenization:**

In [12]:
def tokenize(text:str):
    ''' An example tokenization function. '''

    # simple white-space tokenization:
    return text.lower().split()


In [13]:
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

# Example sentence to encode
s = "Hello, this is a test sentence for the tokenizer."

# Initialize a BPE tokenizer
tokenizer = Tokenizer(BPE())

# Add a pre-tokenizer to handle whitespace properly
tokenizer.pre_tokenizer = Whitespace()

# Initialize a BPE trainer with some default parameters
trainer = BpeTrainer(special_tokens=["<pad>", "<s>", "</s>", "<unk>"])

# Instead of reading files, let's provide training data directly as strings
# Simulate a dataset by giving it a list of strings (or paths to actual files)
training_data = ["Hello world", "This is a test", "We are training the tokenizer"]

# The trainer expects file paths, but we can create files dynamically (or mock them)
# For now, let's assume you have access to these files or use in-memory data.
tokenizer.train_from_iterator(training_data, trainer)

# Encode the string
t = tokenizer.encode(s)

# Decode the tokenized output
decoded_s = tokenizer.decode(t.ids)

# Display the results
print("Original string:", s)
print("Encoded tokens:", t.tokens)
print("Decoded string:", decoded_s)


Original string: Hello, this is a test sentence for the tokenizer.
Encoded tokens: ['Hello', 't', 'h', 'is', 'is', 'a', 'test', 's', 'en', 't', 'en', 'e', 'or', 'the', 'tokenizer']
Decoded string: Hello t h is is a test s en t en e or the tokenizer


**Bag-of-words Embedding:**

See documentation of [sklearn.feature_extraction.text.CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)

In [14]:
from sklearn.feature_extraction.text import CountVectorizer

# create a simple bag of words embedding:
bow = CountVectorizer(

    # the next line converts the filepaths to the actual texts:
    preprocessor = load_text,

    # tokenization function from above:
    tokenizer = tokenize

)

# train the embedding:
embeddings_train = bow.fit_transform(data_train['path'].values)

# vectorize test data:
embeddings_test = bow.transform(data_test['path'].values)

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:521: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


**Classification with a linear SVM**

See documentation of [sklearn.svm.LinearSVC](https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html#sklearn.svm.LinearSVC)

In [15]:
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

svm = LinearSVC()

# train classifier:
svm.fit(embeddings_train, data_train['label'].values)

# test classifier:
predictions = svm.predict(embeddings_test)

# Calculate Accuracy:
print('Accuracy:', accuracy_score(data_test['label'].values, predictions))

Accuracy: 0.766


**TOKENIZATION TASK**

Implement of the chosen tokenizer: BPE
Tokenize a few sample texts
Accuracy score when using BPE tokenization

In [28]:
# Importing necessary libraries
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

# Assuming your dataset 'data_train' and 'data_test' contain paths to text files in the 'path' column
# and the labels in the 'label' column.

# Step 1: Define the function to load the text from file paths
def load_text(filepath):
    with open(filepath, 'r', encoding='utf-8') as file:
        return file.read()

# Step 2: Define the BPE tokenizer and train it on your dataset
def train_bpe_tokenizer(train_texts):
    tokenizer = Tokenizer(BPE())
    trainer = BpeTrainer()
    tokenizer.train_from_iterator(train_texts, trainer=trainer)
    return tokenizer

# Step 3: Tokenize the input text using the trained BPE tokenizer
def tokenize_bpe(tokenizer, text):
    return tokenizer.encode(text).tokens

# Step 4: Create a function to convert texts into embeddings using BPE tokenization
def bpe_vectorize(tokenizer, texts, vectorizer=None): # Add vectorizer argument
    # First, tokenize each text using the BPE tokenizer
    tokenized_texts = [tokenize_bpe(tokenizer, text) for text in texts]

    # Use CountVectorizer to convert tokens to vectors
    # If vectorizer is provided, use it; otherwise, create a new one
    if vectorizer is None:
        vectorizer = CountVectorizer(tokenizer=lambda x: x, preprocessor=None, token_pattern=None, lowercase=False)
        embeddings = vectorizer.fit_transform(tokenized_texts)
    else:
        embeddings = vectorizer.transform(tokenized_texts)  # Use existing vectorizer for transform
    return embeddings, vectorizer # Return the vectorizer as well

# Step 5: Load the actual text content from file paths for both training and test sets
train_texts = [load_text(filepath) for filepath in data_train['path']]
test_texts = [load_text(filepath) for filepath in data_test['path']]

# Step 6: Train the BPE tokenizer on the actual training texts
bpe_tokenizer = train_bpe_tokenizer(train_texts)

# Print a few tokenized sample texts from the dataset
print("Sample Tokenized Texts using BPE Tokenizer:")
for text in train_texts[:5]:  # Display first 5 tokenized texts as an example
    print(tokenize_bpe(bpe_tokenizer, text))

# Step 7: Create embeddings for both training and test sets using BPE tokenizer
embeddings_train_bpe, vectorizer = bpe_vectorize(bpe_tokenizer, train_texts) # Get vectorizer from training
embeddings_test_bpe, _ = bpe_vectorize(bpe_tokenizer, test_texts, vectorizer=vectorizer) # Reuse vectorizer for test

# ... (rest of your code) ...

# Step 8: Train and test a classifier (SVM) on BPE tokenized data
svm = LinearSVC(dual=False, max_iter=5000)

# Train the classifier on the training embeddings
svm.fit(embeddings_train_bpe, data_train['label'].values)

# Test the classifier on the test embeddings
predictions = svm.predict(embeddings_test_bpe)

# Calculate the accuracy on the test set
accuracy_bpe = accuracy_score(data_test['label'].values, predictions)
print(f'Accuracy using BPE Tokenization: {accuracy_bpe:.4f}')


Sample Tokenized Texts using BPE Tokenizer:
['I recently ', 'watched ', 'this', ', but when it ', 'started ', 'I had ', 'no idea ', 'what the ', 'concept ', 'was ', 'about, ', 'what the ', 'topic ', 'was', '.....', 'in ', 'short ', '- ', 'I had ', 'no idea ', 'what it ', 'was. W', 'as ', 'it a ', 'documentar', 'y, was ', 'it a ', 'comedy ', 'rout', 'ine', '.....', 'Well', ', it was ', 'BO', 'TH', '.<br /><br />It ', 'started ', 'a little ', 'slow', ', but ', 'I think', " that's ", 'because I had ', 'absolutely ', 'no idea what ', 'type of ', 'program ', 'I was ', 'viewing', '. But ', 'it ', 'quickly ', 'sucked ', 'me ', 'in. The ', 'episode ', 'I watched ', 'had ', 'Robert ', 'Wu', 'h', 'l ', 'discussing ', 'fact ', 'and ', 'fiction ', 'in ', 'history', '. M', 'ainly ', 'how ', 'we (', 'american', "'s) ", 'learn ', 'history ', 'that ', "isn't really ", 'true ', '- and ', 'how ', 'we ', 'got to ', 'learn ', 'what we ', 'did', '. He ', 'did this ', 'in such a ', 'way as to ', 'keep the '

Noted accuracy is lower with BPE tokenization compared to white space tokenization (0.7270 vs 0.766)

Given above samples of tokenized texts, further text normalization like stemming or custom regular expressions is needed.


First normalization applied without implementing contraction expansion.

In [31]:
# Import necessary libraries
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd
import re
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize
import nltk

# Download NLTK data (if not already downloaded)
nltk.download('punkt')

# Assuming your dataset 'data_train' and 'data_test' contain paths to text files in the 'path' column
# and the labels in the 'label' column.

# Step 1: Define the function to load and normalize the text from file paths
def load_and_normalize_text(filepath):
    with open(filepath, 'r', encoding='utf-8') as file:
        text = file.read()
    # Normalize the text
    text = normalize_text(text)
    return text

# Step 2: Define the normalization function
def normalize_text(text):
    # Convert text to lowercase
    text = text.lower()
    # Remove punctuation using regex
    text = re.sub(r'[^\w\s]', '', text)
    # Optional: Expand contractions (e.g., "don't" to "do not")
    # You can implement contraction expansion here if needed
    # Apply stemming
    stemmer = PorterStemmer()
    tokens = word_tokenize(text)
    stemmed_tokens = [stemmer.stem(token) for token in tokens]
    # Reconstruct the text from stemmed tokens
    normalized_text = ' '.join(stemmed_tokens)
    return normalized_text

# Step 3: Define the BPE tokenizer and train it on your dataset
def train_bpe_tokenizer(train_texts):
    tokenizer = Tokenizer(BPE())
    trainer = BpeTrainer()
    tokenizer.train_from_iterator(train_texts, trainer=trainer)
    return tokenizer

# Step 4: Tokenize the input text using the trained BPE tokenizer
def tokenize_bpe(tokenizer, text):
    return tokenizer.encode(text).tokens

# Step 5: Create a function to convert texts into embeddings using BPE tokenization
def bpe_vectorize(tokenizer, texts):
    # First, tokenize each text using the BPE tokenizer
    tokenized_texts = [tokenize_bpe(tokenizer, text) for text in texts]

    # Use CountVectorizer to convert tokens to vectors
    vectorizer = CountVectorizer(
        tokenizer=lambda x: x,
        preprocessor=None,
        token_pattern=None,
        lowercase=False
    )
    embeddings = vectorizer.fit_transform(tokenized_texts)
    return embeddings, vectorizer

# Step 6: Load and normalize the actual text content from file paths for both training and test sets
train_texts = [load_and_normalize_text(filepath) for filepath in data_train['path']]
test_texts = [load_and_normalize_text(filepath) for filepath in data_test['path']]

# Step 7: Train the BPE tokenizer on the normalized training texts
bpe_tokenizer = train_bpe_tokenizer(train_texts)

# Print a few tokenized sample texts from the dataset
print("Sample Tokenized Texts using BPE Tokenizer with Normalization:")
for text in train_texts[:5]:  # Display first 5 tokenized texts as an example
    print(tokenize_bpe(bpe_tokenizer, text))

# Step 8: Create embeddings for both training and test sets using BPE tokenizer
embeddings_train_bpe_norm, vectorizer = bpe_vectorize(bpe_tokenizer, train_texts)
embeddings_test_bpe_norm = vectorizer.transform(
    [tokenize_bpe(bpe_tokenizer, text) for text in test_texts]
)

# Step 9: Train and test a classifier (SVM) on BPE tokenized data with normalization
svm = LinearSVC(dual=False, max_iter=5000)

# Train the classifier on the training embeddings
svm.fit(embeddings_train_bpe_norm, data_train['label'].values)

# Test the classifier on the test embeddings
predictions = svm.predict(embeddings_test_bpe_norm)

# Calculate the accuracy on the test set
accuracy_bpe_norm = accuracy_score(data_test['label'].values, predictions)
print(f'Accuracy using BPE Tokenization with Normalization: {accuracy_bpe_norm:.4f}')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Sample Tokenized Texts using BPE Tokenizer with Normalization:
['i recent ', 'watch thi ', 'but when it ', 'start ', 'i had no idea ', 'what the ', 'concept ', 'wa about ', 'what the ', 'topic ', 'was', 'in short ', 'i had no idea what ', 'it wa wa ', 'it a ', 'documentari ', 'wa ', 'it a comedi ', 'rout', 'ine', 'wel ', 'it wa ', 'both', 'br br it ', 'start a littl slow ', 'but i ', 'think that ', 'becaus ', 'i had ', 'absolut ', 'no idea what ', 'type of ', 'program ', 'i wa ', 'view ', 'but it ', 'quickli ', 'suck ', 'me ', 'in the ', 'episod ', 'i watch ', 'had ', 'robert ', 'wu', 'hl ', 'discuss ', 'fact ', 'and ', 'fiction ', 'in histori ', 'mainli ', 'how ', 'we ', 'american ', 'learn ', 'histori ', 'that ', 'isnt realli ', 'true ', 'and how ', 'we got ', 'to learn ', 'what we ', 'did ', 'he did ', 'thi ', 'in such a ', 'way as ', 'to keep the viewer ', 'complet ', 'entertain and ', 'interest ', 'i actual ', 'learn ', 'a few ', 'thing ', 'and that ', 'is a true ', 'indic of how 

Noted accuracy is slightly higher with BPE tokenization and some normalization compared to BPE tokenization alone (0.7440 vs 0.7270)


Further text normalization implementing contraction expansion.

In [36]:
# Importing necessary libraries
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import CountVectorizer
import re
from nltk.stem import PorterStemmer
!pip install contractions
import contractions
import nltk

# Ensure NLTK components are downloaded
nltk.download('punkt')

# Step 1: Define the function to load the text from file paths
def load_text(filepath):
    with open(filepath, 'r', encoding='utf-8') as file:
        return file.read()

# Step 2: Define the text normalization function with stemming, regex, and contraction expansion
def normalize_text(text):
    # Expand contractions
    text = contractions.fix(text)  # Expands things like "can't" to "cannot", "I'm" to "I am"

    # Lowercase all text
    text = text.lower()

    # Remove non-alphabetical characters using regex (you can modify this for custom cleaning)
    text = re.sub(r'[^a-zA-Z\s]', '', text)

    # Tokenize words using nltk's word_tokenize
    words = nltk.word_tokenize(text)

    # Apply stemming using PorterStemmer
    stemmer = PorterStemmer()
    stemmed_words = [stemmer.stem(word) for word in words]

    # Join the stemmed words back into a single string
    normalized_text = ' '.join(stemmed_words)

    return normalized_text

# Step 3: Define the BPE tokenizer and train it on your dataset
def train_bpe_tokenizer(train_texts):
    tokenizer = Tokenizer(BPE())
    trainer = BpeTrainer()
    tokenizer.train_from_iterator(train_texts, trainer=trainer)
    return tokenizer

# Step 4: Tokenize the input text using the trained BPE tokenizer
def tokenize_bpe(tokenizer, text):
    return tokenizer.encode(text).tokens

# Step 5: Create a function to convert texts into embeddings using BPE tokenization
def bpe_vectorize(tokenizer, texts, vectorizer=None): # Add vectorizer argument
    # First, tokenize each text using the BPE tokenizer
    tokenized_texts = [tokenize_bpe(tokenizer, text) for text in texts]

    # Use CountVectorizer to convert tokens to vectors
    # If vectorizer is provided, use it; otherwise, create a new one
    if vectorizer is None:
        vectorizer = CountVectorizer(tokenizer=lambda x: x, preprocessor=None, token_pattern=None, lowercase=False)
        embeddings = vectorizer.fit_transform(tokenized_texts)
    else:
        embeddings = vectorizer.transform(tokenized_texts)  # Use existing vectorizer for transform
    return embeddings, vectorizer # Return the vectorizer as well


# Step 6: Load the actual text content from file paths for both training and test sets
train_texts = [load_text(filepath) for filepath in data_train['path']]
test_texts = [load_text(filepath) for filepath in data_test['path']]

# Step 7: Normalize the texts using the custom normalization function (with contraction expansion)
normalized_train_texts = [normalize_text(text) for text in train_texts]
normalized_test_texts = [normalize_text(text) for text in test_texts]

# Step 8: Train the BPE tokenizer on the normalized training data
bpe_tokenizer_normalized = train_bpe_tokenizer(normalized_train_texts)

# Step 9: Print a few tokenized sample texts from the normalized dataset
print("Sample Tokenized Texts using BPE Tokenizer with Normalization:")
for text in normalized_train_texts[:5]:  # Display first 5 tokenized texts as an example
    print(tokenize_bpe(bpe_tokenizer_normalized, text))

# Step 10: Create embeddings for both training and test sets using BPE tokenizer on normalized texts
embeddings_train_bpe_normalized, vectorizer = bpe_vectorize(bpe_tokenizer_normalized, normalized_train_texts) # Get vectorizer from training
embeddings_test_bpe_normalized, _ = bpe_vectorize(bpe_tokenizer_normalized, normalized_test_texts, vectorizer=vectorizer) # Reuse vectorizer for test


# Step 11: Train and test a classifier (SVM) on BPE tokenized normalized data
svm = LinearSVC(dual=False, max_iter=5000)

# Train the classifier on the training embeddings
svm.fit(embeddings_train_bpe_normalized, data_train['label'].values)

# Test the classifier on the test embeddings
predictions_normalized = svm.predict(embeddings_test_bpe_normalized)

# Step 12: Calculate the accuracy on the test set with the additional normalization
accuracy_bpe_normalized = accuracy_score(data_test['label'].values, predictions_normalized)
print(f'Accuracy using BPE Tokenization with Normalization (including contraction expansion): {accuracy_bpe_normalized:.4f}')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Sample Tokenized Texts using BPE Tokenizer with Normalization:
['i recent ', 'watch thi ', 'but when it ', 'start ', 'i had no idea ', 'what the ', 'concept ', 'wa about ', 'what the ', 'topic ', 'was', 'in short ', 'i had no idea ', 'what it wa ', 'wa it a ', 'documentari ', 'wa it a ', 'comedi ', 'routine', 'wel ', 'it wa ', 'both', 'br br it ', 'start a littl slow ', 'but ', 'i think that is ', 'becaus i ', 'had', ' absolut ', 'no idea what ', 'type of program ', 'i wa ', 'view ', 'but it ', 'quickli ', 'suck ', 'me ', 'in the ', 'episod ', 'i watch ', 'had ', 'robert ', 'wu', 'hl ', 'discuss ', 'fact ', 'and ', 'fiction ', 'in histori ', 'mainli ', 'how ', 'we ', 'american ', 'learn ', 'histori', ' that is not ', 'realli ', 'true ', 'and ', 'how ', 'we got ', 'to learn ', 'what we ', 'did he ', 'did thi ', 'in such a ', 'way', ' as to ', 'keep the viewer ', 'complet ', 'entertain and ', 'interest ', 'i actual ', 'learn a ', 'few', ' thing', ' and that ', 'is a true ', 'indic of ', 

Noted accuracy is actually lower with BPE tokenization with Normalization and inclusion of contraction expansion compared to BPE tokenization with Normalization without the contraction expansion (0.7130 vs 0.7440)